f3.go

```diff
@@ -145,412 +78,197 @@ func New(ctx context.Context, manifest manifest.ManifestProvider, ds datastore.D
-	}
-	m.msgSub, err = m.client.topic.Subscribe()
-	if err != nil {
-		m.psLk.Unlock()
-		return xerrors.Errorf("subscribing to topic: %w", err)
-	}
-	m.psLk.Unlock()
-	return nil
-}
+func (m *F3) GetLatestCert(ctx context.Context) (*certs.FinalityCertificate, error) {
+	m.mu.Lock()
+	cs := m.cs
+	m.mu.Unlock()
 
-// Sets up the gpbft runner, this is triggered at initialization
-func (m *F3) startGpbftRunner(ctx context.Context, errCh chan error) {
-	if err := m.initGpbftRunner(ctx); err != nil {
-		errCh <- xerrors.Errorf("initializing gpbft host: %w", err)
-		return
+	if cs == nil {
+		return nil, xerrors.Errorf("F3 is not running")
 	}
-
-	// the size of the buffer is set to prevent message spamming from pubsub,
-	// so it is a big enough buffer to be able to accommodate new messages
-	// at "high-rate", but small enough to avoid spamming or clogging the node.
-	messageQueue := make(chan gpbft.ValidatedMessage, 20)
-	m.client.messageQueue = messageQueue
-
-	go func() {
-		m.mLk.Lock()
-		latest := m.certStore.Latest()
-		m.mLk.Unlock()
-		startInstance := m.Manifest().InitialInstance
-		if latest != nil {
-			startInstance = latest.GPBFTInstance + 1
-		}
-		// Check context before starting the instance
-		select {
-		case <-ctx.Done():
-			// If the context is cancelled before starting, return immediately
-			errCh <- ctx.Err()
-			return
-		default:
-			// Proceed with running the instance
-			err := m.runner.Run(startInstance, ctx)
-			if err != nil {
-				m.log.Errorf("error returned while running host: %+v", err)
-			}
-			errCh <- err
-		}
-	}()
-
-	m.handleIncomingMessages(ctx, messageQueue)
+	return cs.Latest(), nil
 }
 
-func (m *F3) boostrap(ctx context.Context) error {
-	head, err := m.ec.GetHead(ctx)
-	if err != nil {
-		return xerrors.Errorf("failed to get the head: %w", err)
-	}
-
-	if head.Epoch() < m.Manifest().BootstrapEpoch {
-		// wait for bootstrap epoch
-		for {
-			head, err := m.ec.GetHead(ctx)
-			if err != nil {
-				return xerrors.Errorf("getting head: %w", err)
-			}
-			if head.Epoch() >= m.Manifest().BootstrapEpoch {
-				break
-			}
-
-			m.log.Infof("wating for bootstrap epoch (%d): currently at epoch %d", m.Manifest().BootstrapEpoch, head.Epoch())
-			aim := time.Until(head.Timestamp().Add(m.Manifest().ECPeriod))
-			// correct for null epochs
-			for aim < 0 {
-				aim += m.Manifest().ECPeriod
-			}
-
-			select {
-			case <-time.After(aim):
-			case <-ctx.Done():
-				return ctx.Err()
-			}
-		}
-	}
+func (m *F3) GetCert(ctx context.Context, instance uint64) (*certs.FinalityCertificate, error) {
+	m.mu.Lock()
+	cs := m.cs
+	m.mu.Unlock()
 
-	ts, err := m.ec.GetTipsetByEpoch(ctx, m.Manifest().BootstrapEpoch-m.Manifest().ECFinality)
-	if err != nil {
-		return xerrors.Errorf("getting initial power tipset: %w", err)
+	if cs == nil {
+		return nil, xerrors.Errorf("F3 is not running")
 	}
+	return cs.Get(ctx, instance)
+}
 
-	initialPowerTable, err := m.ec.GetPowerTable(ctx, ts.Key())
```

Comment:
This needed to apply the power table delta (that's why I'm now wrapping the `ec.Backend`).